In [95]:
import pandas as pd
import numpy as np

acc_df = pd.read_csv('Accident_Information.csv')

/var/folders/ly/dg_yy0m54vb9my4bfgbtld880000gn/T/ipykernel_27458/3952171714.py:8: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  acc_df = pd.read_csv('Accident_Information.csv')


#### Accident Info

In [96]:
acc_df.head()

Accident_Index 1st_Road_Class  1st_Road_Number 2nd_Road_Class  \
0  200501BS00001              A           3218.0            NaN   
1  200501BS00002              B            450.0              C   
2  200501BS00003              C              0.0            NaN   
3  200501BS00004              A           3220.0            NaN   
4  200501BS00005   Unclassified              0.0            NaN   

   2nd_Road_Number Accident_Severity Carriageway_Hazards        Date  \
0              0.0           Serious                None  2005-01-04   
1              0.0            Slight                None  2005-01-05   
2              0.0            Slight                None  2005-01-06   
3              0.0            Slight                None  2005-01-07   
4              0.0            Slight                None  2005-01-10   

  Day_of_Week  Did_Police_Officer_Attend_Scene_of_Accident  ...  \
0     Tuesday                                          1.0  ...   
1   Wednesday                                          1.0  ...   
2    Thursday                                          1.0  ...   
3      Friday                                          1.0  ...   
4      Monday                                          1.0  ...   

          Police_Force Road_Surface_Conditions           Road_Type  \
0  Metropolitan Police             Wet or damp  Single carriageway   
1  Metropolitan Police                     Dry    Dual carriageway   
2  Metropolitan Police                     Dry  Single carriageway   
3  Metropolitan Police                     Dry  Single carriageway   
4  Metropolitan Police             Wet or damp  Single carriageway   

  Special_Conditions_at_Site Speed_limit   Time  Urban_or_Rural_Area  \
0                       None        30.0  17:42                Urban   
1                       None        30.0  17:36                Urban   
2                       None        30.0  00:15                Urban   
3                       None        30.0  10:35                Urban   
4                       None        30.0  21:13                Urban   

      Weather_Conditions  Year InScotland  
0  Raining no high winds  2005         No  
1     Fine no high winds  2005         No  
2     Fine no high winds  2005         No  
3     Fine no high winds  2005         No  
4     Fine no high winds  2005         No  

[5 rows x 34 columns]

In [97]:
acc_df.shape

(2047256, 34)

In [98]:
acc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2047256 entries, 0 to 2047255
Data columns (total 34 columns):
 #   Column                                       Dtype  
---  ------                                       -----  
 0   Accident_Index                               object 
 1   1st_Road_Class                               object 
 2   1st_Road_Number                              float64
 3   2nd_Road_Class                               object 
 4   2nd_Road_Number                              float64
 5   Accident_Severity                            object 
 6   Carriageway_Hazards                          object 
 7   Date                                         object 
 8   Day_of_Week                                  object 
 9   Did_Police_Officer_Attend_Scene_of_Accident  float64
 10  Junction_Control                             object 
 11  Junction_Detail                              object 
 12  Latitude                                     float64
 13  Light_Condit

In [99]:
acc_df.describe()

1st_Road_Number  2nd_Road_Number  \
count     2.047254e+06     2.029663e+06   
mean      9.921051e+02     3.728153e+02   
std       1.809408e+03     1.287796e+03   
min       0.000000e+00     0.000000e+00   
25%       0.000000e+00     0.000000e+00   
50%       1.180000e+02     0.000000e+00   
75%       7.020000e+02     0.000000e+00   
max       9.999000e+03     9.999000e+03   

       Did_Police_Officer_Attend_Scene_of_Accident      Latitude  \
count                                 2.046978e+06  2.047082e+06   
mean                                  1.202319e+00  5.255978e+01   
std                                   4.081935e-01  1.445506e+00   
min                                   1.000000e+00  4.991294e+01   
25%                                   1.000000e+00  5.148540e+01   
50%                                   1.000000e+00  5.223758e+01   
75%                                   1.000000e+00  5.345590e+01   
max                                   3.000000e+00  6.075754e+01   

       Location_Easting_OSGR  Location_Northing_OSGR     Longitude  \
count           2.047092e+06            2.047092e+06  2.047081e+06   
mean            4.414462e+05            2.968855e+05 -1.410155e+00   
std             9.549620e+04            1.605273e+05  1.403532e+00   
min             6.495000e+04            1.029000e+04 -7.516225e+00   
25%             3.780635e+05            1.777568e+05 -2.329610e+00   
50%             4.430500e+05            2.611835e+05 -1.362233e+00   
75%             5.242982e+05            3.956100e+05 -2.052600e-01   
max             6.555400e+05            1.208800e+06  1.762010e+00   

       Number_of_Casualties  Number_of_Vehicles  \
count          2.047256e+06        2.047256e+06   
mean           1.345843e+00        1.833525e+00   
std            8.179627e-01        7.150543e-01   
min            1.000000e+00        1.000000e+00   
25%            1.000000e+00        1.000000e+00   
50%            1.000000e+00        2.000000e+00   
75%            1.000000e+00        2.000000e+00   
max            9.300000e+01        6.700000e+01   

       Pedestrian_Crossing-Human_Control  \
count                       2.044336e+06   
mean                        1.041707e-02   
std                         1.351126e-01   
min                         0.000000e+00   
25%                         0.000000e+00   
50%                         0.000000e+00   
75%                         0.000000e+00   
max                         2.000000e+00   

       Pedestrian_Crossing-Physical_Facilities   Speed_limit          Year  
count                             2.043696e+06  2.047219e+06  2.047256e+06  
mean                              7.518021e-01  3.884360e+01  2.010524e+03  
std                               1.835289e+00  1.414791e+01  3.765624e+00  
min                               0.000000e+00  0.000000e+00  2.005000e+03  
25%                               0.000000e+00  3.000000e+01  2.007000e+03  
50%                               0.000000e+00  3.000000e+01  2.010000e+03  
75%                               0.000000e+00  5.000000e+01  2.014000e+03  
max                               8.000000e+00  7.000000e+01  2.017000e+03

In [100]:
acc_df.isnull().sum()

Accident_Index                                      0
1st_Road_Class                                      0
1st_Road_Number                                     2
2nd_Road_Class                                 844272
2nd_Road_Number                                 17593
Accident_Severity                                   0
Carriageway_Hazards                                 0
Date                                                0
Day_of_Week                                         0
Did_Police_Officer_Attend_Scene_of_Accident       278
Junction_Control                                    0
Junction_Detail                                     0
Latitude                                          174
Light_Conditions                                    0
Local_Authority_(District)                          0
Local_Authority_(Highway)                           0
Location_Easting_OSGR                             164
Location_Northing_OSGR                            164
Longitude                   

## **1)Handling missing values**

##### **2nd_Road_Class**

Since 2nd road class has 844000 missing values, we can use other data to fill in some of the missing values. We know that if the road number is 0.0 that is mean the class if either unclassified or C. and based on the data in 1st class roads, we know that 30% were unclassified andn 8% were C, so we can use this data to fill in data for 2nd class where the number is 0, substituting in either unlcassified or C randonmly in a 30:8 ratio.


In [101]:
# Create a boolean mask for the rows with missing 2nd_Road_Class and 2nd_Road_Number equals 0
mask = (acc_df['2nd_Road_Class'].isnull()) & (acc_df['2nd_Road_Number'] == 0)

# Calculate the number of rows that meet the condition
num_rows = mask.sum()

# Assign the "C" and "Unclassified" values based on the proportion (21% "C" and 79% "Unclassified")
num_c = int(num_rows * 0.21)
num_unclassified = num_rows - num_c
new_values = ['C'] * num_c + ['Unclassified'] * num_unclassified

# Shuffle the new values to introduce randomness
np.random.shuffle(new_values)

# Update the 2nd_Road_Class values for the masked rows
acc_df.loc[mask, '2nd_Road_Class'] = new_values


In [102]:
print(acc_df['2nd_Road_Class'].isnull().sum())
print(acc_df['2nd_Road_Number'].isnull().sum())

17562
17593


Now that missing values in 2nd class where 2nd road number was 0, we have roughly the same (17000) missing values for each left. This leads me to assume that this is not a mistake in data entry but indicates that there was no second road involved in the accident. So these rows can be substituted in with 'Not Applicable'.

In [103]:
acc_df['2nd_Road_Class'].fillna('Not Applicable', inplace=True)
acc_df['2nd_Road_Number'].fillna('Not Applicable', inplace=True)

##### **Location Columns**

since the latitude and longitude columns provide the same locational data as the easting and northing OSGR, as lat and long is more universally used we can safely drop the easting and northing OSGR columns

In [104]:
acc_df.drop(['Location_Easting_OSGR','Location_Northing_OSGR'], axis=1, inplace=True)

Handling the 144000 missing values from LSOA column:

Initial assumption would be that since there are 144000 misisng values in LSOA and there are 133000 'InScotland' 'Yes' rows that the misisng values mean that they are accidents that took place in scotland. To confirm this we can check how many rows contain both a missing value for LSOA and have 'Yes' for 'InScotland'. 

In [105]:
missing_lsoa_in_scotland = acc_df.loc[acc_df['LSOA_of_Accident_Location'].isna() & (acc_df['InScotland'] == 'Yes')]
print(len(missing_lsoa_in_scotland))

133932


We can see that 133932 rows contain both missing LSOA and Yes inscotland values, confirming that the missing values mean that they are in scotland for the majority. So for these rows we can substitue in 'Scotland'

In [106]:
acc_df.loc[acc_df['InScotland'] == 'Yes', 'LSOA_of_Accident_Location'] = acc_df.loc[acc_df['InScotland'] == 'Yes', 'LSOA_of_Accident_Location'].fillna('SCT')

In [107]:
acc_df['LSOA_of_Accident_Location'].value_counts()

SCT          133932
E01000004      2367
E01004736      1838
E01011365      1503
E01032739      1106
              ...  
E01003044         1
E01029868         1
E01032556         1
E01033717         1
E01032701         1
Name: LSOA_of_Accident_Location, Length: 35565, dtype: int64

##### **Pedestrian_Crossing**


In [108]:
acc_df['Pedestrian_Crossing-Human_Control'] = acc_df['Pedestrian_Crossing-Human_Control'].fillna(9)
acc_df['Pedestrian_Crossing-Physical_Facilities'] = acc_df['Pedestrian_Crossing-Physical_Facilities'].fillna(9)

In [109]:
acc_df.isnull().sum()

Accident_Index                                     0
1st_Road_Class                                     0
1st_Road_Number                                    2
2nd_Road_Class                                     0
2nd_Road_Number                                    0
Accident_Severity                                  0
Carriageway_Hazards                                0
Date                                               0
Day_of_Week                                        0
Did_Police_Officer_Attend_Scene_of_Accident      278
Junction_Control                                   0
Junction_Detail                                    0
Latitude                                         174
Light_Conditions                                   0
Local_Authority_(District)                         0
Local_Authority_(Highway)                          0
Longitude                                        175
LSOA_of_Accident_Location                      11021
Number_of_Casualties                          

##### **1st_Road_Number**
Finding the mode and substituting the 2 missing values with it

In [110]:
mode = acc_df['1st_Road_Number'].mode()[0]
acc_df['1st_Road_Number'] = acc_df['1st_Road_Number'].fillna(mode)

##### **Did_Police_Officer_Attend_Scene_of_Accident**
Finding mode again and imputing it

In [111]:
mode = acc_df['Did_Police_Officer_Attend_Scene_of_Accident'].mode()[0]
acc_df['Did_Police_Officer_Attend_Scene_of_Accident'] = acc_df['Did_Police_Officer_Attend_Scene_of_Accident'].fillna(mode)

##### **Latitude and Longitude**
Creating a dictionary containing the mean or median latitude and longitude values for each Local Authority. Then, for each row with missing latitude and longitude values,use the corresponding Local Authority to fill in the missing values.

In [112]:
# Calculating the mean latitude and longitude for each Local Authority
mean_lat_long_by_local_authority = acc_df.groupby('Local_Authority_(District)').agg({'Latitude': 'mean', 'Longitude': 'mean'}).reset_index()

local_authority_lat_long_dict = dict()

for _, row in mean_lat_long_by_local_authority.iterrows():
    local_authority_lat_long_dict[row['Local_Authority_(District)']] = (row['Latitude'], row['Longitude'])

def fill_missing_lat_long(row, lat_long_dict):
    if pd.isna(row['Latitude']) or pd.isna(row['Longitude']):
        lat, long = lat_long_dict.get(row['Local_Authority_(District)'], (np.nan, np.nan))
        row['Latitude'] = lat
        row['Longitude'] = long
    return row

acc_df = acc_df.apply(lambda row: fill_missing_lat_long(row, local_authority_lat_long_dict), axis=1)

##### **LSOA_of_Accident_Location**
Imputing the missing LSOA values based on the mode of LSOA within the same Local Authority.

In [113]:
mode_lsoa_by_local_authority = acc_df.groupby('Local_Authority_(District)').agg({'LSOA_of_Accident_Location': lambda x: x.mode().iloc[0] if len(x.mode()) > 0 else np.nan}).reset_index()

local_authority_lsoa_dict = dict()

for _, row in mode_lsoa_by_local_authority.iterrows():
    local_authority_lsoa_dict[row['Local_Authority_(District)']] = row['LSOA_of_Accident_Location']

def fill_missing_lsoa(row, lsoa_dict):
    if pd.isna(row['LSOA_of_Accident_Location']):
        lsoa = lsoa_dict.get(row['Local_Authority_(District)'], np.nan)
        row['LSOA_of_Accident_Location'] = lsoa
    return row

acc_df = acc_df.apply(lambda row: fill_missing_lsoa(row, local_authority_lsoa_dict), axis=1)

##### **Speed_limit**
Determining the averge speed (to the nearest 10) of each of the road classes, then assign the speed limt to the rows misisng the values based on what the 1st road class was in that row missing the spoeed limit. 

In [114]:
# calculating mode speed limit for each road class
mode_speed_limit_by_road_class = acc_df.groupby('1st_Road_Class').agg({'Speed_limit': lambda x: x.mode().iloc[0] if len(x.mode()) > 0 else np.nan}).reset_index()
mode_speed_limit_by_road_class

1st_Road_Class  Speed_limit
0              A         30.0
1           A(M)         70.0
2              B         30.0
3              C         30.0
4       Motorway         70.0
5   Unclassified         30.0

In [115]:
# Creating a dictionary containing the mode Speed_limit for each 1st_Road_Class
road_class_speed_limit_dict = dict()

for _, row in mode_speed_limit_by_road_class.iterrows():
    road_class_speed_limit_dict[row['1st_Road_Class']] = row['Speed_limit']

# function to fill missing Speed_limit values based on the 1st_Road_Class
def fill_missing_speed_limit(row, speed_limit_dict):
    if pd.isna(row['Speed_limit']):
        speed_limit = speed_limit_dict.get(row['1st_Road_Class'], np.nan)
        row['Speed_limit'] = speed_limit
    return row

# applying to acc_df
acc_df = acc_df.apply(lambda row: fill_missing_speed_limit(row, road_class_speed_limit_dict), axis=1)

##### **Time**
Randomly assiging time

In [116]:
import random
from datetime import time

def random_time_string():
    random_hour = random.randint(0, 23)
    random_minute = random.randint(0, 59)
    return f"{random_hour:02d}:{random_minute:02d}"

acc_df['Time'] = acc_df['Time'].apply(lambda x: random_time_string() if pd.isna(x) else x)


##### **InScotland**
checking first if the rows in which InScotland is missing that LSOA_of_Accident_Location is present, and if so then it means that it is not in scotland as LSOA only applies to england and wales, there fore nan values can be imputed for 'No'

In [117]:
not_missing_lsoa_in_scotland = acc_df.loc[(acc_df['LSOA_of_Accident_Location'].notna()) & (acc_df['InScotland'].isna())]
print(len(not_missing_lsoa_in_scotland))

53


all 53 of the missing InScotland values occur in rows where LSOA is present, therefore missing values can be changed to 'No'

In [118]:
acc_df['InScotland'].fillna('No', inplace = True)

#### Checking all missing values have been handled...

In [119]:
acc_df.isnull().sum()

Accident_Index                                 0
1st_Road_Class                                 0
1st_Road_Number                                0
2nd_Road_Class                                 0
2nd_Road_Number                                0
Accident_Severity                              0
Carriageway_Hazards                            0
Date                                           0
Day_of_Week                                    0
Did_Police_Officer_Attend_Scene_of_Accident    0
Junction_Control                               0
Junction_Detail                                0
Latitude                                       0
Light_Conditions                               0
Local_Authority_(District)                     0
Local_Authority_(Highway)                      0
Longitude                                      0
LSOA_of_Accident_Location                      0
Number_of_Casualties                           0
Number_of_Vehicles                             0
Pedestrian_Crossing-

All missing values have now successfully been handled without needing to drop any row of data :) 

## **2) Handling Data Types and Formating**

In [120]:
acc_df.dtypes

Accident_Index                                  object
1st_Road_Class                                  object
1st_Road_Number                                float64
2nd_Road_Class                                  object
2nd_Road_Number                                 object
Accident_Severity                               object
Carriageway_Hazards                             object
Date                                            object
Day_of_Week                                     object
Did_Police_Officer_Attend_Scene_of_Accident    float64
Junction_Control                                object
Junction_Detail                                 object
Latitude                                       float64
Light_Conditions                                object
Local_Authority_(District)                      object
Local_Authority_(Highway)                       object
Longitude                                      float64
LSOA_of_Accident_Location                       object
Number_of_

In [121]:
# Convert Date to datetime
acc_df['Date'] = pd.to_datetime(acc_df['Date'])

# Convert 1st_Road_Number and 2nd_Road_Number to int
acc_df['1st_Road_Number'] = acc_df['1st_Road_Number'].astype(int)
# Replace non-numeric values in the '2nd_Road_Number' column with -1
acc_df['2nd_Road_Number'] = pd.to_numeric(acc_df['2nd_Road_Number'], errors='coerce').fillna(-1)
# Convert '2nd_Road_Number' to int
acc_df['2nd_Road_Number'] = acc_df['2nd_Road_Number'].astype(int)

In [122]:
# Convert 'Time' to datetime
acc_df['Time'] = pd.to_datetime(acc_df['Time'], format='%H:%M')

# Extract the time part
acc_df['Time'] = acc_df['Time'].dt.time

In [123]:
def convert_columns_to_categorical(df, columns):
    for column in columns:
        if column in df.columns:
            df[column] = df[column].astype('category')
    return df

# List of columns to convert to 'category' data type
columns_to_convert = [
    '1st_Road_Class',
    '2nd_Road_Class',
    'Accident_Severity',
    'Carriageway_Hazards',
    'Day_of_Week',
    'Did_Police_Officer_Attend_Scene_of_Accident',
    'Junction_Control',
    'Junction_Detail',
    'Light_Conditions',
    'Local_Authority_(District)',
    'Local_Authority_(Highway)',
    'LSOA_of_Accident_Location',
    'Pedestrian_Crossing-Human_Control',
    'Pedestrian_Crossing-Physical_Facilities',
    'Police_Force',
    'Road_Surface_Conditions',
    'Road_Type',
    'Special_Conditions_at_Site',
    'Urban_or_Rural_Area',
    'Weather_Conditions',
    'InScotland'
]

# Convert columns to 'category' data type
acc_df = convert_columns_to_categorical(acc_df, columns_to_convert)

In [124]:
acc_df.dtypes

Accident_Index                                         object
1st_Road_Class                                       category
1st_Road_Number                                         int64
2nd_Road_Class                                       category
2nd_Road_Number                                         int64
Accident_Severity                                    category
Carriageway_Hazards                                  category
Date                                           datetime64[ns]
Day_of_Week                                          category
Did_Police_Officer_Attend_Scene_of_Accident          category
Junction_Control                                     category
Junction_Detail                                      category
Latitude                                              float64
Light_Conditions                                     category
Local_Authority_(District)                           category
Local_Authority_(Highway)                            category
Longitud

In [125]:
acc_df.head()

Accident_Index 1st_Road_Class  1st_Road_Number 2nd_Road_Class  \
0  200501BS00001              A             3218   Unclassified   
1  200501BS00002              B              450              C   
2  200501BS00003              C                0              C   
3  200501BS00004              A             3220   Unclassified   
4  200501BS00005   Unclassified                0   Unclassified   

   2nd_Road_Number Accident_Severity Carriageway_Hazards       Date  \
0                0           Serious                None 2005-01-04   
1                0            Slight                None 2005-01-05   
2                0            Slight                None 2005-01-06   
3                0            Slight                None 2005-01-07   
4                0            Slight                None 2005-01-10   

  Day_of_Week Did_Police_Officer_Attend_Scene_of_Accident  ...  \
0     Tuesday                                         1.0  ...   
1   Wednesday                                         1.0  ...   
2    Thursday                                         1.0  ...   
3      Friday                                         1.0  ...   
4      Monday                                         1.0  ...   

          Police_Force Road_Surface_Conditions           Road_Type  \
0  Metropolitan Police             Wet or damp  Single carriageway   
1  Metropolitan Police                     Dry    Dual carriageway   
2  Metropolitan Police                     Dry  Single carriageway   
3  Metropolitan Police                     Dry  Single carriageway   
4  Metropolitan Police             Wet or damp  Single carriageway   

  Special_Conditions_at_Site Speed_limit      Time  Urban_or_Rural_Area  \
0                       None        30.0  17:42:00                Urban   
1                       None        30.0  17:36:00                Urban   
2                       None        30.0  00:15:00                Urban   
3                       None        30.0  10:35:00                Urban   
4                       None        30.0  21:13:00                Urban   

      Weather_Conditions  Year  InScotland  
0  Raining no high winds  2005          No  
1     Fine no high winds  2005          No  
2     Fine no high winds  2005          No  
3     Fine no high winds  2005          No  
4     Fine no high winds  2005          No  

[5 rows x 32 columns]

In [126]:
acc_df.to_csv('clean_acc_info.csv')